FSharpLu.Json/Compact.fs

```diff
@@ -204,35 +212,57 @@ type CompactUnionJsonConverter(?tupleAsHeterogneousArray:bool) =
                 FSharpValue.MakeUnion(caseSome, [| nestedValue |])
 
         // Tuple type?
-        else if tupleAsHeterogneousArray && isTupleType objectType then
-            if reader.TokenType <> JsonToken.StartArray then
-                failwithf "Expecting a JSON array, got something else"
+        else if isTupleType objectType then
+            if reader.TokenType = JsonToken.StartArray then
+                let tupleType = objectType
+                let elementTypes = FSharpType.GetTupleElements(tupleType)
 
-            // TODO: backward-compat with legacy tuple serialization:
-            // if reader.TokenType is StartObject then we should expecte legacy JSON format for tuples
-            // and fall back on else branch.
+                let readElement elementType =
+                    let more = reader.Read()
+                    if not more then
+                        failwith "Missing array element in deserialized JSON"
 
-            let tupleType = objectType
-            let elementTypes = FSharpType.GetTupleElements(tupleType)
+                    let jToken = Linq.JToken.ReadFrom(reader)
+                    jToken.ToObject(elementType, serializer)
 
-            let readElement elementType =
-                let more = reader.Read()
-                if not more then
-                    failwith "Missing array element in deserialized JSON"
+                let deserializedAsUntypedArray =
+                    elementTypes
+                    |> Array.map readElement
 
-                let jToken = Linq.JToken.ReadFrom(reader)
-                jToken.ToObject(elementType, serializer)
+                let more = reader.Read()
+                if reader.TokenType <> JsonToken.EndArray then
+                    failwith "Expecting end of array token in deserialized JSON"
 
-            let deserializedAsUntypedArray =
-                elementTypes
-                |> Array.map readElement
+                FSharpValue.MakeTuple(deserializedAsUntypedArray, tupleType)
 
-            let more = reader.Read()
-            if reader.TokenType <> JsonToken.EndArray then
-                failwith "Expecting end of array token in deserialized JSON"
+            elif reader.TokenType = JsonToken.StartObject then
+                // if reader.TokenType is StartObject then we should expect legacy JSON format for tuples
 
-            FSharpValue.MakeTuple(deserializedAsUntypedArray, tupleType)
+                let jToken = Linq.JObject.Load(reader)
```

Comment:
Could we fallback on `base.ReadJson(..)` to handle the deserialization from a JSON object?